keps/sig-cluster-lifecycle/kubeadm/4656-add-kubelet-instance-configuration/README.md

```diff
@@ -0,0 +1,287 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The proposal is to remove the Container Runtime Interface (CRI) socket annotation  named `kubeadm.alpha.kubernetes.io/cri-socket` from the Node object in Kubernetes, which is currently added during the `init upload-config` phase in Kubeadm. This annotation is used to specify the CRI socket endpoint used by the kubelet on each node for communication with the container runtime. 
+
+instead of relying on the annotation, this KEP proses creating an instance config per node and overriding the `ContainerRuntimeEndpoint` in the kubelet config when calling kubeadm init and join. This will eliminate the need for kubeadm to store CRI socket configuration on each Node object, instead relying on the kubeadm configuration file to specify this information.
+
+## Motivation
+
+Currently, kubeadm adds a CRI socket annotation to the Node object during the `init upload-config` phase, which specifies the endpoint for the CRI that is being used by the kubelet on each node. This annotation is persistent on the Node object, even if the kubelet is updated or the CRI is changed.
+
+After migrating the container runtime endpoint flag to the instance configuration, we can use it
+Set the CRI socket by overriding the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml`.
+
+### Goals
+
+* kubeadm currently adds an annotation with the key `kubeadm.alpha.kubernetes.io/cri-socket` to each Node object. We will deprecate and remove it.
+* Provide an instance configuration file named `/var/lib/kubelet/instace-config.yaml` for each node, in which the `ContainerRuntimeEndpoint` field is defined. During the `kubeadm init/join/upgrade` process, the instance configuration will be read and the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml` will be overwritten.
+* The `--container-runtime-endpoint` flag is no longer written to the `/var/lib/kubelet/kubeadm-flags.env` file
+
+### Non-Goals
+
+- Continue maintaining CRI socket paths on Node objects.
+
+## Proposal
```

Comment:
the document still does not explain how exactly instance-config.yaml will override config.yaml
i suggest we use patches with the kubeletconfiguration target. this will allow use to override any field.

see here https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/control-plane-flags/

but also the user can pass --paches (or via --config), so the precedence (from high to low) must be:
- user provided kubeletconfig patches with --patches
- instance-config.yaml (uses kubeletconfig patch target to patch config.yaml)
- global config.yaml

WDYT?

if you agree to use the patches approach with kubeletconfig target then the above information must be added under proposal.

Reply:
this part should also start talking about the feature gate.

Reply:
I agree with your opinion, I will add relevant explanations.

Reply:
After carefully reviewing the code, I think we should not directly use the `kubeletconfiguration patch target` mode to implement this feature, which will actually increase the complexity of the implementation. Instead, we should refer to its implementation process and merge `instance-config.yaml` and `kubelet-config.yaml` in the form of calling [ApplyPatchesToTarget](https://github.com/kubernetes/kubernetes/blob/6c556cb9c22d3f4686cbea41e6d301ba889fd6a6/cmd/kubeadm/app/util/patches/patches.go#L160).

Reply:
https://github.com/kubernetes/kubernetes/blob/6c556cb9c22d3f4686cbea41e6d301ba889fd6a6/cmd/kubeadm/app/util/patches/patches.go#L160 does require the target in there to be specified as `kubeletconfiguration` and matching a known target but i think i understand what you mean. and this can be discussed on the k/k PRs too.